NullPointer during Rest client creation when micrometer is disabled

### Describe the bug

NullPointer is thrown if micrometer is included but disabled when RestClient is created.

### Expected behavior

Application to work as usual, but without metrics reported

### Actual behavior

Nullpointer during RestClient creation

```
Caused by: java.lang.NullPointerException
    at io.quarkus.micrometer.runtime.binder.RestClientMetricsListener.prepClientMetrics(RestClientMetricsListener.java:50)
    at io.quarkus.micrometer.runtime.binder.RestClientMetricsListener.onNewClient(RestClientMetricsListener.java:38)
    at io.quarkus.restclient.runtime.QuarkusRestClientBuilder.lambda$build$0(QuarkusRestClientBuilder.java:236)
    at java.base/java.util.ArrayList.forEach(ArrayList.java:1541)
    at io.quarkus.restclient.runtime.QuarkusRestClientBuilder.build(QuarkusRestClientBuilder.java:236)
    at io.quarkus.restclient.runtime.RestClientBase.create(RestClientBase.java:72)
```

### How to Reproduce?

Reproducing project:
https://github.com/danielmaison/quarkus-micrometer-bug

Simple Quarkus project generated with Quarkus 2.8.1.Final and extensions:

* quarkus-resteasy (classic)
* quarkus-resteasy-jsonb
* quarkus-rest-client-jsonb
* quarkus-micrometer-registry-prometheus

**Changes made to sample code**
1. application.properties - disabled micrometer
2. Added Rest client interface with empty get method
3. In Server rest resource, inject rest client

Reproduce by: Calling server resource (e.g. via test)

### Output of `uname -a` or `ver`

Linux Desktop 5.10.60.1-microsoft-standard-WSL2 #1 SMP Wed Aug 25 23:20:18 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux

### Output of `java -version`

openjdk 11.0.14.1 2022-02-08 OpenJDK Runtime Environment (build 11.0.14.1+1-Ubuntu-0ubuntu1.20.04) OpenJDK 64-Bit Server VM (build 11.0.14.1+1-Ubuntu-0ubuntu1.20.04, mixed mode, sharing)

### GraalVM version (if different from Java)

N/A

### Quarkus version or git rev

2.8.1.Final

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

Apache Maven 3.8.4

### Additional information

_No response_

--- a/extensions/micrometer/deployment/src/main/java/io/quarkus/micrometer/deployment/MicrometerProcessor.java
+++ b/extensions/micrometer/deployment/src/main/java/io/quarkus/micrometer/deployment/MicrometerProcessor.java
@@ -75,6 +75,14 @@ public class MicrometerProcessor {
 
     MicrometerConfig mConfig;
 
+    /**
+     * config objects are beans, but they are not unremoveable by default
+     */
+    @BuildStep
+    UnremovableBeanBuildItem mpConfigAsBean() {
+        return UnremovableBeanBuildItem.beanTypes(MicrometerConfig.class);
+    }
+
     @BuildStep(onlyIf = MicrometerEnabled.class)
     FeatureBuildItem feature() {
         return new FeatureBuildItem(Feature.MICROMETER);

--- a/extensions/micrometer/runtime/src/main/java/io/quarkus/micrometer/runtime/binder/RestClientMetricsListener.java
+++ b/extensions/micrometer/runtime/src/main/java/io/quarkus/micrometer/runtime/binder/RestClientMetricsListener.java
@@ -26,13 +26,12 @@ public class RestClientMetricsListener implements RestClientListener {
 
     private final static String REQUEST_METRIC_PROPERTY = "restClientMetrics";
 
-    final MeterRegistry registry = Metrics.globalRegistry;
-    boolean initialized = false;
-    boolean clientMetricsEnabled = false;
+    private final MeterRegistry registry = Metrics.globalRegistry;
+    private boolean initialized = false;
+    private boolean clientMetricsEnabled = false;
 
-    HttpBinderConfiguration httpMetricsConfig;
-    MetricsClientRequestFilter clientRequestFilter;
-    MetricsClientResponseFilter clientResponseFilter;
+    private MetricsClientRequestFilter clientRequestFilter;
+    private MetricsClientResponseFilter clientResponseFilter;
 
     @Override
     public void onNewClient(Class<?> serviceInterface, RestClientBuilder builder) {
@@ -51,11 +50,11 @@ public class RestClientMetricsListener implements RestClientListener {
     boolean prepClientMetrics() {
         boolean clientMetricsEnabled = this.clientMetricsEnabled;
         if (!this.initialized) {
-            this.httpMetricsConfig = Arc.container().instance(HttpBinderConfiguration.class).get();
+            HttpBinderConfiguration httpMetricsConfig = Arc.container().instance(HttpBinderConfiguration.class).get();
             clientMetricsEnabled = httpMetricsConfig.isClientEnabled();
             if (clientMetricsEnabled) {
-                this.clientRequestFilter = new MetricsClientRequestFilter(httpMetricsConfig);
-                this.clientResponseFilter = new MetricsClientResponseFilter();
+                this.clientRequestFilter = new MetricsClientRequestFilter(registry);
+                this.clientResponseFilter = new MetricsClientResponseFilter(registry, httpMetricsConfig);
             }
             this.clientMetricsEnabled = clientMetricsEnabled;
             this.initialized = true;
@@ -63,11 +62,11 @@ public class RestClientMetricsListener implements RestClientListener {
         return clientMetricsEnabled;
     }
 
-    class MetricsClientRequestFilter implements ClientRequestFilter {
-        HttpBinderConfiguration binderConfiguration;
+    static class MetricsClientRequestFilter implements ClientRequestFilter {
+        private final MeterRegistry registry;
 
-        MetricsClientRequestFilter(HttpBinderConfiguration binderConfiguration) {
-            this.binderConfiguration = binderConfiguration;
+        MetricsClientRequestFilter(MeterRegistry registry) {
+            this.registry = registry;
         }
 
         @Override
@@ -78,7 +77,16 @@ public class RestClientMetricsListener implements RestClientListener {
         }
     }
 
-    class MetricsClientResponseFilter implements ClientResponseFilter {
+    static class MetricsClientResponseFilter implements ClientResponseFilter {
+        private final MeterRegistry registry;
+        private final HttpBinderConfiguration httpMetricsConfig;
+
+        MetricsClientResponseFilter(MeterRegistry registry,
+                HttpBinderConfiguration httpMetricsConfig) {
+            this.registry = registry;
+            this.httpMetricsConfig = httpMetricsConfig;
+        }
+
         @Override
         public void filter(ClientRequestContext requestContext, ClientResponseContext responseContext) throws IOException {
             RequestMetricInfo requestMetric = getRequestMetric(requestContext);
@@ -120,7 +128,7 @@ public class RestClientMetricsListener implements RestClientListener {
         }
     }
 
-    class RestClientMetricInfo extends RequestMetricInfo {
+    static class RestClientMetricInfo extends RequestMetricInfo {
         ClientRequestContext requestContext;
 
         RestClientMetricInfo(ClientRequestContext requestContext) {

--- a/extensions/micrometer/runtime/src/main/java/io/quarkus/micrometer/runtime/binder/RestClientMetricsListener.java
+++ b/extensions/micrometer/runtime/src/main/java/io/quarkus/micrometer/runtime/binder/RestClientMetricsListener.java
@@ -17,6 +17,7 @@ import io.micrometer.core.instrument.Tag;
 import io.micrometer.core.instrument.Tags;
 import io.micrometer.core.instrument.Timer;
 import io.quarkus.arc.Arc;
+import io.quarkus.micrometer.runtime.config.MicrometerConfig;
 
 /**
  * This is initialized via ServiceFactory (static/non-CDI initialization)
@@ -35,6 +36,10 @@ public class RestClientMetricsListener implements RestClientListener {
 
     @Override
     public void onNewClient(Class<?> serviceInterface, RestClientBuilder builder) {
+        MicrometerConfig micrometerConfig = Arc.container().instance(MicrometerConfig.class).get();
+        if (!micrometerConfig.enabled) {
+            return;
+        }
         if (prepClientMetrics()) {
             // This must run AFTER the OpenTelmetry client request filter
             builder.register(this.clientRequestFilter, Priorities.HEADER_DECORATOR + 1);

